// @blockworks-foundation/mango-client/lib/src/commands/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.sanityCheck = exports.listMarket = exports.setStubOracle = exports.initGroup = exports.addSwitchboardOracle = exports.addPythOracle = exports.addStubOracle = exports.addSpotMarket = exports.addPerpMarket = void 0;
const initGroup_1 = __importDefault(require("./initGroup"));
exports.initGroup = initGroup_1.default;
const addPerpMarket_1 = __importDefault(require("./addPerpMarket"));
exports.addPerpMarket = addPerpMarket_1.default;
const addSpotMarket_1 = __importDefault(require("./addSpotMarket"));
exports.addSpotMarket = addSpotMarket_1.default;
const addStubOracle_1 = __importDefault(require("./addStubOracle"));
exports.addStubOracle = addStubOracle_1.default;
const addPythOracle_1 = __importDefault(require("./addPythOracle"));
exports.addPythOracle = addPythOracle_1.default;
const addSwitchboardOracle_1 = __importDefault(require("./addSwitchboardOracle"));
exports.addSwitchboardOracle = addSwitchboardOracle_1.default;
const setStubOracle_1 = __importDefault(require("./setStubOracle"));
exports.setStubOracle = setStubOracle_1.default;
const listMarket_1 = __importDefault(require("./listMarket"));
exports.listMarket = listMarket_1.default;
const sanityCheck_1 = __importDefault(require("./sanityCheck"));
exports.sanityCheck = sanityCheck_1.default;
